### ISSUE TITLE
Using System.Reflection.TypeExtensions.GetProperties in .NET Core 2.0

### ISSUE THREAD

**Issue by Stamo-Gochev** [dotnet/corefx#28100]
I am using a Class Library project, whose target frameworks are `net451` and `netstandard1.6`, which dynamically compiles a class in a way similar to the example here (a "standard" way to do the compilation):

https://github.com/dotnet/roslyn/issues/16211#issuecomment-372998985

but inside a method in `MyClass`, I want to retrieve all public properties like this (the actual code is the string of this, which is passed to `CSharpSyntaxTree.ParseText`:
```csharp
using System;
public class MyClass
{
    public static void Print()
    {
#if NET451
            var props = this.GetType().GetProperties(System.Reflection.BindingFlags.Instance | System.Reflection.BindingFlags.Public);
#else
            var props = System.Reflection.TypeExtensions.GetProperties(this.GetType(), System.Reflection.BindingFlags.Instance | System.Reflection.BindingFlags.Public);
#endif
             //additional code
    }   
}
```
Now, the Class Library project is referenced by an ASP.NET Core web app (assume any third party app that I have no control over), which targets `netcore1.1` and as [`System.Reflection.TypeExtensions.GetProperties` is available](https://docs.microsoft.com/en-us/dotnet/api/system.reflection.typeextensions.getproperties?view=netcore-1.1#System_Reflection_TypeExtensions_GetProperties_System_Type_System_Reflection_BindingFlags_), it is working as expected.

However, targeting `netcore2.0` throws an error although it looks like the [method should still be available](https://docs.microsoft.com/en-us/dotnet/api/system.reflection.typeextensions.getproperties?view=netcore-2.0#System_Reflection_TypeExtensions_GetProperties_System_Type_System_Reflection_BindingFlags_). The alternative (used for `net451` above):
```csharp
var props = this.GetType().GetProperties(System.Reflection.BindingFlags.Instance | System.Reflection.BindingFlags.Public);
```
is actually working for `netcore2.0`, but not for `netcore1.0` or `netcore1.1`.

Can you provide more information why this is happening and if it is required for the `netcore2.0` project to "provide" the `System.Reflection.TypeExtensions` package?

**Comment by danmosemsft** [comment]
@weshaggard seems this is because of your change below?

commit a4364b283016758aff4d8ea162e6e8c550e2da6f
Author: Wes Haggard <Wes.Haggard@microsoft.com>
Date:   Mon Oct 16 11:46:43 2017 -0700

    Update SupportedFrameworks to netcoreapp2.1 for some packages

**Comment by weshaggard** [comment]
@danmosemsft I don't believe it is related to that commit. It most likely has to do with changes in the shared framework between 1.1 and 2.0

@Stamo-Gochev is is kind of hard to follow exactly what you are doing and what is compiled when. Could you setup a repo with some sample projects that demonstrate the issue you are hitting and I will take a look and help you figure out a solution.

**Comment by Stamo-Gochev** [comment]
@danmosemsft @weshaggard 

I will be working on isolating the issue in a project that I can send you as the project is not open-sourced, but what I have found so far is that the problem is caused by the `System.Reflection.TypeExtensions` assembly not being referenced when the compilation is done and thus the `GetProperties` method is not available.

The references to the assemblies that are necessary for the compilation are obtained by the `GetReferences` method below - the Class Library project uses the [HostingEnvironment](https://docs.microsoft.com/en-us/dotnet/api/microsoft.aspnetcore.hosting.ihostingenvironment?view=aspnetcore-2.0) variable, which is passed to it from the ASP.NET Core web app:

```csharp
public class MyLibrary {
    //injected by the ASP.NET Core web app
    private readonly IHostingEnvironment _environment;

    ...
    public void Compile() 
    {
        //assume a correct string definition of the class
        var tree = CSharpSyntaxTree.ParseText(@"
        using System;
        public class MyClass
        {
            public static void Main()
            {
                var instance = new MyClass();
                var props = System.Reflection.TypeExtensions.GetProperties(instance.GetType(), System.Reflection.BindingFlags.Instance | System.Reflection.BindingFlags.Public);
            }   
        }");
        ...
        var compilation = CSharpCompilation.Create(assemblyName,
                        options: new CSharpCompilationOptions(OutputKind.DynamicallyLinkedLibrary),
                        syntaxTrees: new[] { tree },
                        references: GetReferences()

        ...
    }
    ...

    //get the avaialble references based on the HostingEnvironment variable
    public IEnumerable<MetadataReference> GetReferences()
    {
        var references = new List<MetadataReference>();

        var applicationAssembly = Assembly.Load(new AssemblyName(_environment.ApplicationName));
        var context = DependencyContext.Load(applicationAssembly);      

        if (context != null)
        {
            for (var i = 0; i < context.CompileLibraries.Count; i++)
            {
                var library = context.CompileLibraries[i];
                IEnumerable<string> referencePaths;
                try
                {
                    referencePaths = library.ResolveReferencePaths();
                }
                catch (InvalidOperationException e)
                {
                    continue;
                }

                //add the references
                references.AddRange(referencePaths...);
            }
        }

        return references;
    }
    ...
}
```

The above code works fine in ASP.NET Core 1.1 project and the full list of loaded assemblies can be found at: https://pastebin.com/NBevj597
The `System.Reflection.TypeExtensions` assembly is loaded and thus the compilation is successful.

On the other hande, the same code fails when executed in ASP.NET Core 2.0 project: https://pastebin.com/abETUZ8E
The `System.Reflection.TypeExtensions` assembly is not loaded. In addition, `InvalidOperationException`s are caught for the following entries:
```
"Cannot find compilation library location for package 'Microsoft.Win32.Registry'"
"Cannot find compilation library location for package 'System.Data.SqlClient'"
"Cannot find compilation library location for package 'System.Runtime.CompilerServices.Unsafe'"
"Cannot find compilation library location for package 'System.Security.AccessControl'"
"Cannot find compilation library location for package 'System.Security.Cryptography.Xml'"
"Cannot find compilation library location for package 'System.Security.Principal.Windows'"
"Cannot find compilation library location for package 'Microsoft.NETCore.App'"
```

I assume the whole thing is dependent on what is exposed by the `HostingEnvironment` variable in the above snippet:
```csharp
var applicationAssembly = Assembly.Load(new AssemblyName(_environment.ApplicationName));
var context = DependencyContext.Load(applicationAssembly);

(context.CompileLibraries)
```
After the above observations, it looks like the issue is not connected to `dotnet/corefx` responsibilities at all, so it should be closed, but can you forward it to somebody that can provide more information about the differences between `netcore1.1` and `netcore2.0` (probably the ASP.NET team)?

**Comment by weshaggard** [comment]
@davidfowl @anurse any ideas why HostingEnvironment seems to not work here? Is there some difference between 1.1 and 2.0 that needs to be accounted for?

**Comment by AtsushiKan** [comment]
Reassigning as right now, the investigation actionability is in ASP.NET, not CoreFx.

**Comment by AtsushiKan** [comment]
This should be closed and forwarded to ASP.NET but I don't know which repo to do that in. Changing the milestone so it stops showing up a 2.1 ZBB blocker in CoreFx.

**Comment by danmosemsft** [comment]
@davidfowl where should this move to

**Comment by anurse** [comment]
HostingEnvironment is at https://github.com/aspnet/Hosting.
